### DISCUSSION TITLE
GraphQL union and useQuery reexecution w/ graphcache

### DISCUSSION THREAD

**Discussion by nephix** [urql-graphql/urql#2917] [Q&A] [answered]
Hey, under what circumstances does useQuery reexecute and gives me the latest cache entry?

I'm running into a query with GraphQL union types where I don't get the latest cache entry

Query:
```graphql
query GetCurrentUser {
  currentUser {
    ... on User {
      __typename
      id
      profile {
        name
      }
      isAnonymous
    }
    ... on Anonymous {
      __typename
      isAnonymous
    }
  }
}
```

Mutation:
```graphql
mutation UpdateUserProfile($payload: CurrentUserProfileUpdate!) {
  updateCurrentUserProfile(payload: $payload) {
    user {
      id
      profile {
        name
      }
      isAnonymous
    }
  }
}
```

After the mutation executed, I would expect the query above to reexecute and deliver the latest cache entry:

```typescript
const [{ data }] = useGetCurrentUserQuery(); // generated hook, uses Urql.useQuery
```

But the data in `data` is always outdated after I ran the mutation.

I also introduced fragments and added mutation updater to the urql client. 

Inside the updater function, I used `cache.readFragment` to get the latest cache entry for the user. 

It shows me the *most recent* data returned from the mutation, so `user` returned by the `updateCurrentUserProfile` is cached.

My questions:

1. Why does `useGetCurrentUserQuery` not return the latest cache entry?
2. Is it a problem caused by the `Anonymous` type not having an `id`/`key` field?
3. Why is the `data` of `useGetCurrentUserQuery` updated when I refetch via `const [{ data }, refetch] = useGetCurrentUserQuery()` but not when a mutation updates an existing cache entry?

**Comment by kitten** [comment, marked as answer]
There's a small nuance here of "links", which are explained here: https://formidable.com/open-source/urql/docs/graphcache/normalized-caching/

Basically, you have a root field called `Query.currentUser`. This field may start out pointing at `Anonymous` without an ID, or to a `User` with an ID. In other words, you have a "link" (or relation) that's pointing from `Query.currentUser` to another entity.

When `Mutation.updateCurrentUserProfile.user` updates an entity, this doesn't give Graphcache any indication that `Query.currentUser` may have changed. In other words, because you're not using `Query.currentUser`, that field hasn't changed.

So, this is _not_ caused by `Anonymous` not having an ID. Rather, it's caused by that field not being mentioned in the mutation. This therefore requires a manual update. Such as:

```js
cacheExchange({
  updates: {
    updateCurrentUserProfile(result) {
      cache.link('Query', 'currentUser', result.updateCurrentUserProfile.user);
    },
  }
})
```

This API is further explained here: https://formidable.com/open-source/urql/docs/graphcache/cache-updates/#writing-links-individually
It basically manually writes a link to the cache during an update.

**Reply by nephix** [reply]
Makes sense, thanks 🙏
